lib/ProductOpener/KnowledgePanels.pm

```diff
@@ -885,4 +887,116 @@ sub create_ingredients_panel($$$) {
         $panel_data_ref, $product_ref, $target_lc, $target_cc);
 }
 
+
+=head2 create_additives_panel ( $product_ref, $target_lc, $target_cc )
+
+Creates knowledge panels for additives.
+
+=head3 Arguments
+
+=head4 product reference $product_ref
+
+=head4 language code $target_lc
+
+=head4 country code $target_cc
+
+=cut
+
+sub create_additives_panel($$$) {
+
+	my $product_ref = shift;
+	my $target_lc = shift;
+	my $target_cc = shift;
+
+	$log->debug("create additives panel", { code => $product_ref->{code} }) if $log->is_debug();
+
+    # Create a panel only if the product has additives
+
+    if ((defined $product_ref->{additives_tags}) and (scalar @{$product_ref->{additives_tags}} > 0 )) {
+
+        my $additives_panel_data_ref = {
+        };
+
+        foreach my $additive (@{$product_ref->{additives_tags}}) {
+
+            my $additive_panel_id = "additive_" . $additive;
+
+            my $additive_panel_data_ref = {
+                additive => $additive,
+            };
+
+            # Wikipedia abstracts, in target language or English
+
+            my $target_lcs_ref = [$target_lc];
+            if ($target_lc ne "en") {
+                push @$target_lcs_ref, "en";
+            }
+
+            add_taxonomy_properties_in_target_languages_to_object($additive_panel_data_ref, "additives", $additive,
+                ["wikipedia_url", "wikipedia_title", "wikipedia_abstract"], $target_lcs_ref);
+
+            create_panel_from_json_template("additive_" . $additive, "api/knowledge-panels/health/ingredients/additive.tt.json",
+                $additive_panel_data_ref, $product_ref, $target_lc, $target_cc);
+        }
+
+        create_panel_from_json_template("additives", "api/knowledge-panels/health/ingredients/additives.tt.json",
+            $additives_panel_data_ref, $product_ref, $target_lc, $target_cc);
+
+    }
+}
+
+
+=head2 add_taxonomy_properties_in_target_languages_to_object ( $object_ref, $tagtype, $tagid, $properties_ref, $target_lcs_ref )
+
+This function adds to the hash ref $object_ref (for instance a data structure passed to a template) the values
+of selected properties, if they exist in one of the target languages.
+
+For instance for the panel for an additive, we can include a Wikipedia abstract in the requested language if available,
+or in English if not.
+
+=head3 Arguments
+
+=head4 object reference $object_ref
+
+=head4 taxonomy $tagtype
+
+=head4 tag id $tagoid
+
+=head4 list of properties $properties_ref
+
+Properties to add to the resulting object.
+
+=head4 language codes $target_lcs
+
+Reference to an array of preferred languages, with the preferred language first.
+
+=cut
+
+sub add_taxonomy_properties_in_target_languages_to_object ($$$$$) {
+
+    my $object_ref = shift;
+    my $tagtype = shift;
+    my $tagid = shift;
+
+    my $properties_ref = shift;
+    my $target_lcs_ref = shift;
```

Comment:
When there is more than 3 or 4 arguments, shouldn't we use hash to pass arguments ?

I feel like this could help make the code a bit more readable (where we call such functions).

Reply:
If we use a hash, we lose the check on the number of arguments that Perl does at compile time. There isn't a very good way to have signatures in Perl unfortunately (there's an experimental "use experimental 'signatures';" and some modules).